From 0157322d79712fd6e5c758029459d978883b311c Mon Sep 17 00:00:00 2001
From: Marcelo Leitner <mleitner@redhat.com>
Date: Tue, 21 Oct 2014 16:16:44 -0400
Subject: [net] netfilter: nf_tables: restore notifications for anonymous set destruction

Message-id: <20141021161644.22888.17874.stgit@localhost.localdomain>
Patchwork-id: 97924
O-Subject: [RHEL7 net PATCH 024/124] netfilter: nf_tables: restore notifications for anonymous set destruction
Bugzilla: 1148041 1155088
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>

Backport of upstream commit: ab9da5c19f359f9ac2635157d9cd45deec4ef63c

 commit ab9da5c19f359f9ac2635157d9cd45deec4ef63c
 Author: Patrick McHardy <kaber@trash.net>
 Date:   Fri Mar 7 19:08:31 2014 +0100

    netfilter: nf_tables: restore notifications for anonymous set destruction

    Since we have the context available again, we can restore notifications
    for destruction of anonymous sets.

    Signed-off-by: Patrick McHardy <kaber@trash.net>
    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Marcelo Ricardo Leitner <mleitner@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/netfilter/nf_tables_api.c b/net/netfilter/nf_tables_api.c
index 2c10c3f..33045a5 100644
--- a/net/netfilter/nf_tables_api.c
+++ b/net/netfilter/nf_tables_api.c
@@ -2442,8 +2442,7 @@ err1:
 static void nf_tables_set_destroy(const struct nft_ctx *ctx, struct nft_set *set)
 {
  list_del(&set->list);
- if (!(set->flags & NFT_SET_ANONYMOUS))
-  nf_tables_set_notify(ctx, set, NFT_MSG_DELSET);
+ nf_tables_set_notify(ctx, set, NFT_MSG_DELSET);
 
  set->ops->destroy(set);
  module_put(set->ops->owner);
diff --git a/net/netfilter/nft_lookup.c b/net/netfilter/nft_lookup.c
index 953978e..7fd2bea 100644
--- a/net/netfilter/nft_lookup.c
+++ b/net/netfilter/nft_lookup.c
@@ -94,7 +94,7 @@ static void nft_lookup_destroy(const struct nft_ctx *ctx,
 {
  struct nft_lookup *priv = nft_expr_priv(expr);
 
- nf_tables_unbind_set(NULL, priv->set, &priv->binding);
+ nf_tables_unbind_set(ctx, priv->set, &priv->binding);
 }
 
 static int nft_lookup_dump(struct sk_buff *skb, const struct nft_expr *expr)
-- 
1.7.1